Allow resetting/cleaning PlatformIO resources

avilcheslopez opened the pull request:
- Fixes #2403

cmcantalupo committed:
Add a reset method for PlatformIO

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

avilcheslopez committed:
Add python binding for geopm_pio_reset()

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

cmcantalupo commented on service/src/PlatformIO.cpp:
```diff
@@ -27,10 +28,24 @@
+
+    static PlatformIO &platform_io_helper(bool do_reset)
+    {
+        static std::unique_ptr<PlatformIOImp> instance;
```
```suggestion
        static std::unique_ptr<PlatformIOImp> instance = geopm::make_unique<PlatformIOImp>();
```

cmcantalupo committed:
Fix static constructor for PlatformIO instance

- Initialized as nullptr otherwise

Signed-off-by: Christopher M. Cantalupo <christpher.m.cantalupo@intel.com>

cmcantalupo commented:
There are documentation changes required for all the PIO interfaces for push_signal(), and push_control().  These have language bindings for C, C++, python.  For all of these there is user documentation and developer documentation (man pages, doxygen strings, and python docstrings).  The C++ interface register_iogroup() should also be updated about the effect of reset.

Some examples that should reflect the effect of reset:

After a reset the stack will be gone, and the user will be able to create a new stack:
https://github.com/geopm/geopm/blob/dev/service/src/geopm_pio.h#L144-L153

Any IOGroup that was registered by a user (outside of the PlatformIO constructor) prior to a reset will need to be registered again:
https://github.com/geopm/geopm/blob/dev/service/src/geopm/PlatformIO.hpp#L36-L44

avilcheslopez committed:
Update documentation for PIO APIs affected by reset

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add clarification to geopm_pio_reset() documentation

The reset only applies to the Service PlatformIO instance and not the
instance managed by the runtime.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add integration test for PlatformIO reset functionality

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>
